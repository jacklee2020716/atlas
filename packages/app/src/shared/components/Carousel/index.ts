import Carousel, { CarouselProps } from './Carousel'
import { CAROUSEL_CONTROL_SIZE } from './Carousel.style'

export { Carousel as default, CAROUSEL_CONTROL_SIZE }
export type { CarouselProps }
