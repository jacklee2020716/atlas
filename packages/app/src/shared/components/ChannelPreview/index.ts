import ChannelPreview from './ChannelPreview'
import ChannelPreviewBase from './ChannelPreviewBase'

export { ChannelPreview, ChannelPreviewBase }
