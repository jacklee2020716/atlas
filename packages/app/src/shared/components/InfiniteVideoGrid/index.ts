import InfiniteVideoGrid from './InfiniteVideoGrid'
export default InfiniteVideoGrid
