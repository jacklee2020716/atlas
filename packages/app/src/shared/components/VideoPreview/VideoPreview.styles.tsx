import styled from '@emotion/styled'
import { colors, spacing, typography } from '../../theme'
import Avatar from '../Avatar'
import { PlayIcon } from '../../icons'

const HOVER_BORDER_SIZE = '2px'

type CoverImageProps = {
  displayPosterPlaceholder?: boolean
}

type ContainerProps = {
  clickable: boolean
}

type ChannelProps = {
  channelClickable: boolean
}

export const CoverContainer = styled.div`
  width: 320px;
  height: 190px;

  transition-property: box-shadow, transform;
  transition-duration: 0.4s;
  transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1);

  position: relative;
`

export const CoverImage = styled.img<CoverImageProps>`
  width: 100%;
  height: 100%;
  background-image: ${({ displayPosterPlaceholder }) =>
    displayPosterPlaceholder ? `linear-gradient(${colors.gray[300]}, ${colors.gray[700]})` : 'none'};
  background-size: cover;
  object-fit: cover;
`

export const CoverHoverOverlay = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  opacity: 0;

  transition: opacity 0.4s cubic-bezier(0.165, 0.84, 0.44, 1);

  border: ${HOVER_BORDER_SIZE} solid ${colors.white};
  background: linear-gradient(180deg, #000000 0%, rgba(0, 0, 0, 0) 100%);

  display: flex;
  justify-content: center;
  align-items: center;
`

export const CoverPlayIcon = styled(PlayIcon)`
  transition: transform 0.4s cubic-bezier(0.165, 0.84, 0.44, 1);
  transform: translateY(40px);
  width: 54px;
  height: 54px;
`

export const ProgressOverlay = styled.div`
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  height: ${spacing.xxs};
  background-color: ${colors.white};
`

export const ProgressBar = styled.div`
  position: absolute;
  left: 0;
  bottom: 0;
  height: 100%;
  width: 0;
  background-color: ${colors.blue['500']};
`

export const Container = styled.div<ContainerProps>`
  color: ${colors.gray[300]};
  cursor: ${({ clickable }) => (clickable ? 'pointer' : 'auto')};
  display: inline-block;
  ${({ clickable }) =>
    clickable &&
    `
				&:hover {
					${CoverContainer} {
						transform: translate(-${spacing.xs}, -${spacing.xs});
						box-shadow: ${spacing.xs} ${spacing.xs} 0 ${colors.blue['500']};
					}

					${CoverHoverOverlay} {
						opacity: 1;
					}
					
					${CoverPlayIcon} {
						transform: translateY(0);
					}

					${ProgressOverlay} {
						bottom: ${HOVER_BORDER_SIZE};
					}
				}
			`}
`

export const CoverDurationOverlay = styled.div`
  position: absolute;
  bottom: ${spacing.xs};
  right: ${spacing.xs};
  padding: ${spacing.xxs} ${spacing.xs};
  background-color: ${colors.black};
  color: ${colors.white};
  font-size: ${typography.sizes.body2};
`

export const InfoContainer = styled.div`
  display: flex;
  margin-top: ${spacing.s};
`

export const StyledAvatar = styled(Avatar)<ChannelProps>`
  width: 40px;
  height: 40px;
  margin-right: ${spacing.xs};
  cursor: ${({ channelClickable }) => (channelClickable ? 'pointer' : 'auto')};
`

export const TextContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: start;
`

export const TitleHeader = styled.h3`
  margin: 0;
  font-weight: ${typography.weights.bold};
  font-size: ${typography.sizes.h6};
  line-height: 1.25rem;
  color: ${colors.white};
  display: inline-block;
`

export const ChannelName = styled.span<ChannelProps>`
  font-size: ${typography.sizes.subtitle2};
  line-height: 1.25rem;
  display: inline-block;
  cursor: ${({ channelClickable }) => (channelClickable ? 'pointer' : 'auto')};
`

export const MetaText = styled.span`
  margin-top: ${spacing.xs};
  font-size: ${typography.sizes.subtitle2};
`
