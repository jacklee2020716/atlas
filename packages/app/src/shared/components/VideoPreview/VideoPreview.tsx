import React from 'react'
import {
  ChannelName,
  CoverDurationOverlay,
  CoverHoverOverlay,
  CoverImage,
  CoverPlayIcon,
  MetaText,
  ProgressBar,
  ProgressOverlay,
  StyledAvatar,
  TitleHeader,
} from './VideoPreview.styles'
import { formatDateAgo, formatDurationShort } from '@/utils/time'
import { formatNumberShort } from '@/utils/number'
import VideoPreviewBase from './VideoPreviewBase'

type VideoPreviewProps = {
  title: string
  channelName: string
  channelAvatarURL?: string | null
  createdAt: Date
  duration?: number
  // video watch progress in percent (0-100)
  progress?: number
  views: number
  posterURL: string

  showChannel?: boolean
  showMeta?: boolean
  imgRef?: React.Ref<HTMLImageElement>
  onClick?: (e: React.MouseEvent<HTMLElement>) => void
  onChannelClick?: (e: React.MouseEvent<HTMLElement>) => void
  className?: string
}

const VideoPreview: React.FC<VideoPreviewProps> = ({
  title,
  channelName,
  channelAvatarURL,
  createdAt,
  duration,
  progress = 0,
  views,
  posterURL,
  showChannel = true,
  showMeta = true,
  imgRef,
  onClick,
  onChannelClick,
  className,
}) => {
  const channelClickable = !!onChannelClick

  const handleChannelClick = (e: React.MouseEvent<HTMLElement>) => {
    if (!onChannelClick) {
      return
    }
    e.stopPropagation()
    onChannelClick(e)
  }

  const handleClick = (e: React.MouseEvent<HTMLElement>) => {
    if (!onClick) {
      return
    }
    e.stopPropagation()
    onClick(e)
  }

  const coverNode = (
    <>
      <CoverImage src={posterURL} ref={imgRef} alt={`${title} by ${channelName} thumbnail`} />
      {duration && <CoverDurationOverlay>{formatDurationShort(duration)}</CoverDurationOverlay>}
      {!!progress && (
        <ProgressOverlay>
          <ProgressBar style={{ width: `${progress}%` }} />
        </ProgressOverlay>
      )}
      <CoverHoverOverlay>
        <CoverPlayIcon />
      </CoverHoverOverlay>
    </>
  )

  const titleNode = <TitleHeader>{title}</TitleHeader>

  const channelAvatarNode = (
    <StyledAvatar
      size="small"
      name={channelName}
      img={channelAvatarURL}
      channelClickable={channelClickable}
      onClick={handleChannelClick}
    />
  )

  const channelNameNode = (
    <ChannelName channelClickable={channelClickable} onClick={handleChannelClick}>
      {channelName}
    </ChannelName>
  )

  const metaNode = (
    <MetaText>
      {formatDateAgo(createdAt)}ãƒ»{formatNumberShort(views)} views
    </MetaText>
  )

  return (
    <VideoPreviewBase
      coverNode={coverNode}
      titleNode={titleNode}
      showChannel={showChannel}
      channelAvatarNode={channelAvatarNode}
      channelNameNode={channelNameNode}
      showMeta={showMeta}
      metaNode={metaNode}
      onClick={handleClick}
      className={className}
    />
  )
}

export default VideoPreview
