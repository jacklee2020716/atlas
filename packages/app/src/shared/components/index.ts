export { default as Avatar } from "./Avatar";
export { default as Button } from "./Button";
export { default as Carousel } from "./Carousel";
export { default as Dropdown } from "./Dropdown";
export { default as Grid } from "./Grid";
export { default as Header } from "./Header";
export { default as Link } from "./Link";
export { default as NavButton } from "./NavButton";
export { default as RadioButton } from "./RadioButton";
export { default as Checkbox } from "./Checkbox";
export { default as TagButton } from "./TagButton";
export { default as Tabs } from "./Tabs";
export { default as Tab } from "./Tabs/Tab";
export { default as Tag } from "./Tag";
export { default as TextField } from "./TextField";
export { default as Typography } from "./Typography";
export { default as VideoPreview } from "./VideoPreview";
export { default as VideoPlayer } from "./VideoPlayer";
export { default as SeriesPreview } from "./SeriesPreview";
export { default as ChannelPreview } from "./ChannelPreview";
export { default as HamburgerButton } from "./HamburgerButton";
export { default as Gallery } from "./Gallery";
export { default as Sidenav, SIDENAV_WIDTH, EXPANDED_SIDENAV_WIDTH, NavItem } from "./Sidenav";
export { default as GlobalStyle } from "./GlobalStyle";
