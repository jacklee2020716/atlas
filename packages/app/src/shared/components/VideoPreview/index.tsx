import VideoPreview from './VideoPreview'
import VideoPreviewBase from './VideoPreviewBase'
import { MAX_VIDEO_PREVIEW_WIDTH } from './VideoPreviewBase.styles'

export { VideoPreview, VideoPreviewBase, MAX_VIDEO_PREVIEW_WIDTH }
