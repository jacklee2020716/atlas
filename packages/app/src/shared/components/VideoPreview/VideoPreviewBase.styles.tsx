import styled from '@emotion/styled'
import { colors, spacing } from '@/shared/theme'
import { CoverHoverOverlay, CoverPlayIcon, ProgressOverlay } from './VideoPreview.styles'

export const HOVER_BORDER_SIZE = '2px'

type ContainerProps = {
  clickable: boolean
}

export const MAX_VIDEO_PREVIEW_WIDTH = '320px'

export const CoverContainer = styled.div`
  width: 100%;
  height: 190px;

  transition-property: box-shadow, transform;
  transition-duration: 0.4s;
  transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1);

  position: relative;
`

export const Container = styled.article<ContainerProps>`
  max-width: ${MAX_VIDEO_PREVIEW_WIDTH};
  color: ${colors.gray[300]};
  cursor: ${({ clickable }) => (clickable ? 'pointer' : 'auto')};
  display: inline-block;
  ${({ clickable }) =>
    clickable &&
    `
				&:hover {
					${CoverContainer} {
						transform: translate(-${spacing.xs}, -${spacing.xs});
						box-shadow: ${spacing.xs} ${spacing.xs} 0 ${colors.blue['500']};
					}

					${CoverHoverOverlay} {
						opacity: 1;
					}
					
					${CoverPlayIcon} {
						transform: translateY(0);
					}

					${ProgressOverlay} {
						bottom: ${HOVER_BORDER_SIZE};
					}
				}
			`}
`

export const InfoContainer = styled.div`
  display: flex;
  margin-top: ${spacing.s};
`

export const AvatarContainer = styled.div`
  width: 40px;
  min-width: 40px;
  height: 40px;
  margin-right: ${spacing.xs};
`

export const TextContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: start;
  width: 100%;
`

export const MetaContainer = styled.div`
  margin-top: ${spacing.xs};
  width: 100%;
`
