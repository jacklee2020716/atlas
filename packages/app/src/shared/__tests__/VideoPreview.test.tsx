import React from 'react'
import { mount } from 'enzyme'
import { VideoPreview } from '@/shared/components/VideoPreview'
describe('VideoPreview component', () => {
  it('Should render Video Preview correctly', () => {
    expect(
      mount(
        <VideoPreview
          title="Some Video Title"
          channelName="some channel"
          posterURL=""
          views={1000}
          createdAt={new Date()}
        />
      )
    ).toBeDefined()
  })
})
