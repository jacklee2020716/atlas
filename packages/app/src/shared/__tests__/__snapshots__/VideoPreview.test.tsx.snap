// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPreview component Should render Video Preview correctly 1`] = `ReactWrapper {}`;
